pkg/core/backoff.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2018 Pusher Ltd. and Wave Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// MinBackoff is the minimum backoff duration (1 second)
+	MinBackoff = 1 * time.Second
+	// MaxBackoff is the maximum backoff duration (64 seconds)
+	MaxBackoff = 64 * time.Second
+)
+
+// backoffState tracks the backoff state for a single instance
+type backoffState struct {
+	lastUpdateTime time.Time
+	backoffLevel   int // 0 = 1s, 1 = 2s, 2 = 4s, 3 = 8s, 4 = 16s, 5 = 32s, 6 = 64s
+}
+
+// BackoffTracker manages exponential backoff state for multiple instances
+type BackoffTracker struct {
+	states map[types.NamespacedName]*backoffState
+	mutex  sync.RWMutex
+}
+
+// NewBackoffTracker creates a new BackoffTracker
+func NewBackoffTracker() *BackoffTracker {
+	return &BackoffTracker{
+		states: make(map[types.NamespacedName]*backoffState),
+	}
+}
+
+// CalculateBackoff returns the backoff duration for the given level
+func CalculateBackoff(level int) time.Duration {
+	// Cap level to prevent overflow (level 6 = 64s which is our max)
+	if level > 6 {
+		level = 6
+	}
+	backoff := MinBackoff * (1 << level) // 2^level seconds
+	if backoff > MaxBackoff {
+		backoff = MaxBackoff
+	}
+	return backoff
+}
+
+// ShouldUpdate checks if an instance should be updated based on backoff state
+// Returns (shouldUpdate, requeueAfter)
+func (bt *BackoffTracker) ShouldUpdate(name types.NamespacedName) (bool, time.Duration) {
+	bt.mutex.Lock()
+	defer bt.mutex.Unlock()
+
+	state, exists := bt.states[name]
+	now := time.Now()
+
+	// First update or no backoff state
+	if !exists {
+		bt.states[name] = &backoffState{
+			lastUpdateTime: now,
+			backoffLevel:   0,
+		}
+		return true, 0
+	}
+
+	// Check if enough time has passed since last update
+	elapsed := now.Sub(state.lastUpdateTime)
+	backoff := CalculateBackoff(state.backoffLevel)
+
+	if elapsed < backoff {
+		// Still in backoff period - delay update
+		remaining := backoff - elapsed
+		return false, remaining
+	}
+
+	// Backoff period has passed, allow update and increase level
+	state.lastUpdateTime = now
+	state.backoffLevel++
+	// Cap at level 6 (64s)
+	if state.backoffLevel > 6 {
+		state.backoffLevel = 6
```

Comment:
This is redundant to the MaxBackoff check above. Either way works but both together do not make it better.